magma/smart/smart_bits.py

```diff
@@ -0,0 +1,526 @@
+        self._width_ = to_width
+        self._signed_ = all(arg._signed_ for arg in self._args)
+
+    def eval(self):
+        args = self._eval_args()
+        if self._signed_:
+            args = (sint(arg) for arg in args)
+        else:
+            args = (uint(arg) for arg in args)
+        return self.op(*args)
+
+
+class _SmartBinaryOpExpr(_SmartNAryContextualOpExr):
+    def __init__(self, op, loperand, roperand):
+        super().__init__(op, loperand, roperand)
+
+
+class _SmartUnaryOpExpr(_SmartNAryContextualOpExr):
+    def __init__(self, op, operand):
+        super().__init__(op, operand)
+
+
+class _SmartComparisonOpExpr(_SmartOpExpr):
+    def __init__(self, op, loperand, roperand):
+        super().__init__(op, loperand, roperand)
+
+    def resolve(self, context):
+        context = Context(None, self)
+        self._resolve_args(context)
+        to_width = max(arg._width_ for arg in self._args)
+        args = (_extend_if_needed(arg, to_width) for arg in self._args)
+        self._update(*args)
+        self._width_ = 1
+        self._signed_ = False
+
+    def eval(self):
+        args = self._eval_args()
+        signed_args = all(arg._signed_ for arg in self._args)
+        fn = sint if signed_args else uint
+        args = (fn(arg) for arg in args)
+        return uint(self.op(*args))
+
+
+class _SmartShiftOpExpr(_SmartOpExpr):
+    def __init__(self, op, loperand, roperand):
+        super().__init__(op, loperand, roperand)
+
+    def resolve(self, context):
+        loperand, roperand = self._args
+        loperand.resolve(context)
+        self_context = Context(None, self)
+        roperand.resolve(self_context)
+        to_width = max(arg._width_ for arg in self._args)
+        args = (_extend_if_needed(arg, to_width) for arg in self._args)
+        self._update(*args)
+        self._width_ = to_width
+        self._signed_ = all(arg._signed_ for arg in self._args)
+
+    def eval(self):
+        args = self._eval_args()
+        if self._signed_:
+            args = (sint(arg) for arg in args)
+        else:
+            args = (uint(arg) for arg in args)
+        return self.op(*args)
+
+
+class _SmartConcatOpExpr(_SmartOpExpr):
+
+    class _ConcatOp:
+        def __call__(self, *args):
+            return bits_concat(*args)
+
+        def __str__(self):
+            return "Concat"
+
+    def __init__(self, *args):
+        concat = _SmartConcatOpExpr._ConcatOp()
+        super().__init__(concat, *args)
+
+    def resolve(self, context):
+        for arg in self._args:
+            context = Context(None, arg)
+            arg.resolve(context)
+        self._width_ = sum(arg._width_ for arg in self._args)
+        self._signed_ = False
+
+    def eval(self):
+        args = self._eval_args()
+        args = [uint(arg) for arg in args]
+        return uint(self.op(*args))
+
+
+def concat(*args):
+    if not all(isinstance(arg, _SmartExpr) for arg in args):
+        return NotImplemented
```

Comment:
Maybe this should raise NotImplementedError instead?

Reply:
Agreed.  `NotImplemented` is used for operators to signal that the lhs doesn't know how to handle the rhs and is converted to a type error if the rhs also does not how handle the left.   It should not be returned to a user.

Reply:
ah ok missed that subtlety. thanks. fixed.